Associated items support

```rust
struct Foo;

impl Foo {
    const BAR: i32 = 42;
}

fn main() {
    let x = Foo::BAR;
}
```

`Foo::BAR` does not get resolved.

I tried to look into it, but for some reason I have trouble building the code after making some changes.

```
$ cargo install-code
[snip]
$ git stash pop
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   complete_path.rs

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (0d39549b2709395fc5569444efdb9ab2b919202c)
$ cargo install-code # why is it rebuilding the tools?
   Compiling num-traits v0.2.6
   Compiling serde v1.0.87
   Compiling spin v0.4.10
   Compiling syn v0.15.26
   Compiling lazy_static v1.2.0
   Compiling thread_local v0.3.6
   Compiling regex v1.1.0
   Compiling num-integer v0.1.39
   Compiling serde_derive v1.0.87
   Compiling pest_generator v2.1.0
   Compiling synstructure v0.10.1
   Compiling pest_derive v2.1.0
   Compiling failure_derive v0.1.5
   Compiling failure v0.1.5
   Compiling chrono v0.4.6
   Compiling serde_json v1.0.38
   Compiling ron v0.4.1
   Compiling tera v0.11.20
   Compiling teraron v0.0.1
   Compiling tools v0.1.0 (/home/grayshade/rust-analyzer/crates/tools)
    Finished dev [unoptimized + debuginfo] target(s) in 32.41s
     Running `/home/grayshade/rust-analyzer/target/debug/tools install-code`

will run: cargo install --path crates/ra_lsp_server --force
  Installing ra_lsp_server v0.1.0 (/home/grayshade/rust-analyzer/crates/ra_lsp_server)
   Compiling ra_ide_api v0.1.0 (/home/grayshade/rust-analyzer/crates/ra_ide_api)
error[E0599]: no method named `parse` found for type `&db::RootDatabase` in the current scope
  --> crates/ra_ide_api/src/completion.rs:49:28
   |
49 |     let original_file = db.parse(position.file_id);
   |                            ^^^^^
   |
   = note: the method `parse` exists but the following trait bounds were not satisfied:
           `&db::RootDatabase : ra_db::SourceDatabase`
           `db::RootDatabase : ra_db::SourceDatabase`
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `parse`, perhaps you need to implement it:
           candidate #1: `ra_db::SourceDatabase`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: failed to compile `ra_lsp_server v0.1.0 (/home/grayshade/rust-analyzer/crates/ra_lsp_server)`, intermediate artifacts can be found at `/home/grayshade/rust-analyzer/target`

Caused by:
  Could not compile `ra_ide_api`.

To learn more, run the command again with --verbose.
Error: ErrorMessage { msg: "`cargo install --path crates/ra_lsp_server --force` exited with exit code: 101" }
$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   complete_path.rs

no changes added to commit (use "git add" and/or "git commit -a")
$ git diff                 
diff --git i/crates/ra_ide_api/src/completion/complete_path.rs w/crates/ra_ide_api/src/completion/complete_path.rs
index 39aefdb1..e05b4eff 100644
--- i/crates/ra_ide_api/src/completion/complete_path.rs
+++ w/crates/ra_ide_api/src/completion/complete_path.rs
@@ -7,14 +7,17 @@ use crate::{
 };
 
 pub(super) fn complete_path(acc: &mut Completions, ctx: &CompletionContext) {
+    log::info!("complete_path");
     let path = match &ctx.path_prefix {
         Some(path) => path.clone(),
         _ => return,
     };
+    log::info!("{:?}", path);
     let def = match ctx.resolver.resolve_path(ctx.db, &path).take_types() {
         Some(Resolution::Def(def)) => def,
         _ => return,
     };
+    log::info!("{:?}", def);
     match def {
         hir::ModuleDef::Module(module) => {
             let module_scope = module.scope(ctx.db);
@@ -48,6 +51,10 @@ pub(super) fn complete_path(acc: &mut Completions, ctx: &CompletionContext) {
                 }
             });
         }
+        // hir::ModuleDef::Struct(s) => {
+        //     let ty = s.ty(ctx.db);
+        //     log::info!("{:?}", ty);
+        // }
         _ => return,
     };
 }
```

Notice that I haven't touched the file it complains about.

--- a/crates/ra_hir/src/ty/method_resolution.rs
+++ b/crates/ra_hir/src/ty/method_resolution.rs
@@ -174,4 +174,24 @@ impl Ty {
         }
         None
     }
+
+    // This would be nicer if it just returned an iterator, but that runs into
+    // lifetime problems, because we need to borrow temp `CrateImplBlocks`.
+    pub fn iterate_impl_items<T>(
+        self,
+        db: &impl HirDatabase,
+        mut callback: impl FnMut(ImplItem) -> Option<T>,
+    ) -> Option<T> {
+        let krate = def_crate(db, &self)?;
+        let impls = db.impls_in_crate(krate);
+
+        for (_, impl_block) in impls.lookup_impl_blocks(db, &self) {
+            for item in impl_block.items() {
+                if let Some(result) = callback(*item) {
+                    return Some(result);
+                }
+            }
+        }
+        None
+    }
 }

--- a/crates/ra_ide_api/src/completion.rs
+++ b/crates/ra_ide_api/src/completion.rs
@@ -80,3 +80,25 @@ pub fn function_label(node: &ast::FnDef) -> Option<String> {
 
     Some(label.trim().to_owned())
 }
+
+pub fn const_label(node: &ast::ConstDef) -> String {
+    let label: String = node
+        .syntax()
+        .children()
+        .filter(|child| ast::Comment::cast(child).is_none())
+        .map(|node| node.text().to_string())
+        .collect();
+
+    label.trim().to_owned()
+}
+
+pub fn type_label(node: &ast::TypeDef) -> String {
+    let label: String = node
+        .syntax()
+        .children()
+        .filter(|child| ast::Comment::cast(child).is_none())
+        .map(|node| node.text().to_string())
+        .collect();
+
+    label.trim().to_owned()
+}

--- a/crates/ra_ide_api/src/completion/complete_path.rs
+++ b/crates/ra_ide_api/src/completion/complete_path.rs
@@ -1,6 +1,6 @@
 use join_to_string::join;
 use hir::{Docs, Resolution};
-use ra_syntax::AstNode;
+use ra_syntax::{AstNode, ast::NameOwner};
 use test_utils::tested_by;
 
 use crate::completion::{CompletionItem, CompletionItemKind, Completions, CompletionKind, CompletionContext};
@@ -55,6 +55,51 @@ pub(super) fn complete_path(acc: &mut Completions, ctx: &CompletionContext) {
                     .set_documentation(variant.docs(ctx.db))
                     .set_detail(Some(detail))
                     .add_to(acc)
+                }
+            });
+        }
+        hir::ModuleDef::Struct(s) => {
+            let ty = s.ty(ctx.db);
+            ty.iterate_impl_items(ctx.db, |item| match item {
+                hir::ImplItem::Method(func) => {
+                    let sig = func.signature(ctx.db);
+                    if !sig.has_self_param() {
+                        CompletionItem::new(
+                            CompletionKind::Reference,
+                            ctx.source_range(),
+                            sig.name().to_string(),
+                        )
+                        .from_function(ctx, func)
+                        .kind(CompletionItemKind::Method)
+                        .add_to(acc);
+                    }
+                    None::<()>
+                }
+                hir::ImplItem::Const(ct) => {
+                    let source = ct.source(ctx.db);
+                    if let Some(name) = source.1.name() {
+                        CompletionItem::new(
+                            CompletionKind::Reference,
+                            ctx.source_range(),
+                            name.text().to_string(),
+                        )
+                        .from_const(ctx, ct)
+                        .add_to(acc);
+                    }
+                    None::<()>
+                }
+                hir::ImplItem::Type(ty) => {
+                    let source = ty.source(ctx.db);
+                    if let Some(name) = source.1.name() {
+                        CompletionItem::new(
+                            CompletionKind::Reference,
+                            ctx.source_range(),
+                            name.text().to_string(),
+                        )
+                        .from_type(ctx, ty)
+                        .add_to(acc);
+                    }
+                    None::<()>
                 }
             });
         }
@@ -193,6 +238,63 @@ mod tests {
                 S(S),
             }
             fn foo() { let _ = E::<|> }
+            ",
+        );
+    }
+
+    #[test]
+    fn completes_struct_associated_method() {
+        check_reference_completion(
+            "struct_associated_method",
+            "
+            //- /lib.rs
+            /// A Struct
+            struct S;
+
+            impl S {
+                /// An associated method
+                fn m() { }
+            }
+
+            fn foo() { let _ = S::<|> }
+            ",
+        );
+    }
+
+    #[test]
+    fn completes_struct_associated_const() {
+        check_reference_completion(
+            "struct_associated_const",
+            "
+            //- /lib.rs
+            /// A Struct
+            struct S;
+
+            impl S {
+                /// An associated const
+                const C: i32 = 42;
+            }
+
+            fn foo() { let _ = S::<|> }
+            ",
+        );
+    }
+
+    #[test]
+    fn completes_struct_associated_type() {
+        check_reference_completion(
+            "struct_associated_type",
+            "
+            //- /lib.rs
+            /// A Struct
+            struct S;
+
+            impl S {
+                /// An associated type
+                type T = i32;
+            }
+
+            fn foo() { let _ = S::<|> }
             ",
         );
     }

--- a/crates/ra_ide_api/src/completion/completion_item.rs
+++ b/crates/ra_ide_api/src/completion/completion_item.rs
@@ -8,6 +8,8 @@ use test_utils::tested_by;
 use crate::completion::{
     completion_context::CompletionContext,
     function_label,
+    const_label,
+    type_label
 };
 
 /// `CompletionItem` describes a single completion variant in the editor pop-up.
@@ -267,6 +269,28 @@ impl Builder {
         self.kind = Some(CompletionItemKind::Function);
         self
     }
+
+    pub(super) fn from_const(mut self, ctx: &CompletionContext, ct: hir::Const) -> Builder {
+        if let Some(docs) = ct.docs(ctx.db) {
+            self.documentation = Some(docs);
+        }
+
+        self.detail = Some(const_item_label(ctx, ct));
+        self.kind = Some(CompletionItemKind::Const);
+
+        self
+    }
+
+    pub(super) fn from_type(mut self, ctx: &CompletionContext, ty: hir::Type) -> Builder {
+        if let Some(docs) = ty.docs(ctx.db) {
+            self.documentation = Some(docs);
+        }
+
+        self.detail = Some(type_item_label(ctx, ty));
+        self.kind = Some(CompletionItemKind::TypeAlias);
+
+        self
+    }
 }
 
 impl<'a> Into<CompletionItem> for Builder {
@@ -303,6 +327,16 @@ impl Into<Vec<CompletionItem>> for Completions {
 fn function_item_label(ctx: &CompletionContext, function: hir::Function) -> Option<String> {
     let node = function.source(ctx.db).1;
     function_label(&node)
+}
+
+fn const_item_label(ctx: &CompletionContext, ct: hir::Const) -> String {
+    let node = ct.source(ctx.db).1;
+    const_label(&node)
+}
+
+fn type_item_label(ctx: &CompletionContext, ty: hir::Type) -> String {
+    let node = ty.source(ctx.db).1;
+    type_label(&node)
 }
 
 #[cfg(test)]

